File: server/lib/lark.ts
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';

const run = promisify(execFile);

const DEFAULT_OPTS = {
  maxBuffer: 64 * 1024 * 1024,
  timeout: 60_000,
} as const;

export interface LarkChat {
  chat_id: string;
  name: string;
  description?: string;
  chat_type?: string;
  chat_mode?: string;
  user_count?: string;
  add_member_permission?: string;
  // CLI pretty/table output may use camelCase or snake_case
  memberCount?: number;
  chatId?: string;
}

export interface LarkSender {
  id: string;
  id_type?: string;
  name?: string;
  // pretty output sometimes flattens
  sender_id?: string;
  sender_name?: string;
}

export interface LarkMessage {
  message_id: string;
  chat_id?: string;
  sender?: LarkSender;
  create_time?: string; // ms timestamp as string
  update_time?: string;
  msg_type?: string;
  content?: string; // JSON string for text/post/card
  body?: {
    content?: string;
  };
  mentions?: Array<{
    key: string;
    id: { open_id: string; user_id?: string };
    name: string;
    tenant_key?: string;
  }>;
  parent_id?: string;
  thread_id?: string;
  // pretty output may flatten
  messageId?: string;
  msgType?: string;
  createTime?: string;
}

export interface LarkChatListResponse {
  ok: boolean;
  data?: {
    items?: LarkChat[];
    chats?: LarkChat[];
    has_more?: boolean;
    page_token?: string;
  };
  chats?: LarkChat[];
  error?: { type: string; message: string };
}

export interface LarkMessagesResponse {
  ok: boolean;
  data?: {
    items?: LarkMessage[];
    messages?: LarkMessage[];
    has_more?: boolean;
    page_token?: string;
  };
  messages?: LarkMessage[];
  error?: { type: string; message: string };
}

async function larkRaw(args: string[], opts = DEFAULT_OPTS): Promise<string> {
  const { stdout } = await run('lark-cli', args, opts);
  return stdout;
}

async function larkJson<T>(args: string[], opts = DEFAULT_OPTS): Promise<T> {
  const stdout = await larkRaw([...args, '--json'], opts);
  return JSON.parse(stdout) as T;
}

export async function larkAvailable(): Promise<boolean> {
  try {
    await run('lark-cli', ['--version'], { timeout: 5_000 });
    return true;
  } catch {
    return false;
  }
}

export async function larkDoctor(): Promise<{
  ok: boolean;
  configured: boolean;
  authenticated: boolean;
  identity?: string;
  error?: string;
}> {
  try {
    const out = await larkRaw(['doctor']);
    const parsed = JSON.parse(out) as {
      ok?: boolean;
      checks?: Array<{ name: string; status: string; message?: string }>;
    };
    const checks = parsed.checks ?? [];
    const configOk = checks.some((c) => c.name === 'config_file' && c.status === 'pass');
    const appOk = checks.some((c) => c.name === 'app_resolved' && c.status === 'pass');
    const userOk = checks.some((c) => c.name === 'user_identity' && c.status === 'pass');
    return {
      ok: parsed.ok ?? false,
      configured: configOk && appOk,
      authenticated: userOk,
      error: parsed.ok ? undefined : out.slice(0, 200),
    };
  } catch (e) {
    return {
      ok: false,
      configured: false,
      authenticated: false,
      error: e instanceof Error ? e.message : 'unknown',
    };
  }
}

export async function larkChatList(pageToken?: string, pageSize = 100): Promise<LarkChat[]> {
  const args = ['im', '+chat-list', '--as', 'user', '--page-size', String(pageSize)];
  if (pageToken) args.push('--page-token', pageToken);
  const res = await larkJson<LarkChatListResponse>(args);
  if (!res.ok && res.error) {
    throw new Error(`lark chat-list failed: ${res.error.type} ${res.error.message}`);
  }
  const items = (res.data?.items ?? res.data?.chats ?? res.chats ?? []) as LarkChat[];
  // Normalize field names
  return items.map((c) => ({
    chat_id: c.chat_id || c.chatId || '',
    name: c.name || '未命名群',
    description: c.description,
    chat_type: c.chat_type || c.chat_mode,
    user_count: c.user_count,
    memberCount: c.memberCount,
  }));
}

export async function larkChatMessages(
  chatId: string,
  opts: {
    start?: string; // ISO8601
    end?: string; // ISO8601
    pageToken?: string;
    pageSize?: number;
    sort?: 'asc' | 'desc';
  } = {},
): Promise<{ messages: LarkMessage[]; hasMore: boolean; pageToken?: string }> {
  const args = [
    'im',
    '+chat-messages-list',
    '--as',
    'user',
    '--chat-id',
    chatId,
    '--page-size',
    String(opts.pageSize ?? 50),
    '--sort',
    opts.sort ?? 'desc',
    '--no-reactions',
  ];
  if (opts.start) args.push('--start', opts.start);
  if (opts.end) args.push('--end', opts.end);
  if (opts.pageToken) args.push('--page-token', opts.pageToken);

  const res = await larkJson<LarkMessagesResponse>(args);
  if (!res.ok && res.error) {
    throw new Error(`lark chat-messages-list failed: ${res.error.type} ${res.error.message}`);
  }
  const items = (res.data?.items ?? res.data?.messages ?? res.messages ?? []) as LarkMessage[];
  return {
    messages: items.map(normalizeMessage),
    hasMore: res.data?.has_more ?? false,
    pageToken: res.data?.page_token,
  };
}

function normalizeMessage(m: LarkMessage): LarkMessage {
  return {
    ...m,
    message_id: m.message_id || m.messageId || '',
    msg_type: m.msg_type || m.msgType || 'unknown',
    create_time: m.create_time || m.createTime,
    sender: m.sender
      ? {
          id: m.sender.id || (m.sender as unknown as { sender_id?: string }).sender_id || '',
          name: m.sender.name || (m.sender as unknown as { sender_name?: string }).sender_name || '',
        }
      : {
          id: (m as unknown as { sender_id?: string }).sender_id || '',
          name: (m as unknown as { sender_name?: string }).sender_name || '',
        },
  };
}

export async function larkAllChats(): Promise<LarkChat[]> {
  const out: LarkChat[] = [];
  let pageToken: string | undefined;
  do {
    const batch = await larkChatList(pageToken, 100);
    out.push(...batch);
    pageToken = batch.length === 100 ? undefined : undefined; // lark-cli shortcut may not expose page_token; stop after one batch for MVP
  } while (pageToken);
  return out;
}

export async function larkAllMessages(
  chatId: string,
  opts: { start?: string; end?: string; maxPages?: number } = {},
): Promise<LarkMessage[]> {
  const out: LarkMessage[] = [];
  let pageToken: string | undefined;
  let pages = 0;
  do {
    const batch = await larkChatMessages(chatId, {
      ...opts,
      pageToken,
      pageSize: 50,
      sort: 'desc',
    });
    out.push(...batch.messages);
    pageToken = batch.pageToken;
    pages++;
  } while (pageToken && pages < (opts.maxPages ?? 4));
  return out;
}

File: apps/data-service/models/lark.go
package models

// LarkChat represents a Lark chatroom
type LarkChat struct {
	ChatID      string `json:"chat_id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	ChatType    string `json:"chat_type,omitempty"`
	ChatMode    string `json:"chat_mode,omitempty"`
	UserCount   string `json:"user_count,omitempty"`
	MemberCount int    `json:"memberCount,omitempty"`
	ChatIdAlt   string `json:"chatId,omitempty"`
}

// LarkSender represents a message sender
type LarkSender struct {
	ID         string `json:"id"`
	IDType     string `json:"id_type,omitempty"`
	Name       string `json:"name,omitempty"`
	SenderID   string `json:"sender_id,omitempty"`
	SenderName string `json:"sender_name,omitempty"`
}

// LarkMessage represents a Lark message
type LarkMessage struct {
	MessageID   string `json:"message_id"`
	ChatID      string `json:"chat_id,omitempty"`
	Sender      *LarkSender `json:"sender,omitempty"`
	CreateTime  string `json:"create_time,omitempty"`
	UpdateTime  string `json:"update_time,omitempty"`
	MsgType     string `json:"msg_type,omitempty"`
	Content     string `json:"content,omitempty"`
	Body        *LarkMessageBody `json:"body,omitempty"`
	Mentions    []LarkMention `json:"mentions,omitempty"`
	ParentID    string `json:"parent_id,omitempty"`
	ThreadID    string `json:"thread_id,omitempty"`
	MessageIdAlt string `json:"messageId,omitempty"`
	MsgTypeAlt  string `json:"msgType,omitempty"`
	CreateTimeAlt string `json:"createTime,omitempty"`
}

// LarkMessageBody represents the message body
type LarkMessageBody struct {
	Content string `json:"content,omitempty"`
}

// LarkMention represents a mention in a message
type LarkMention struct {
	Key         string `json:"key"`
	ID          string `json:"id"`
	Name        string `json:"name"`
	TenantKey   string `json:"tenant_key,omitempty"`
}

// LarkChatListResponse represents the chat list API response
type LarkChatListResponse struct {
	OK    bool `json:"ok"`
	Data  *LarkChatListData `json:"data,omitempty"`
	Chats []LarkChat `json:"chats,omitempty"`
	Error *LarkError `json:"error,omitempty"`
}

// LarkChatListData represents the chat list data
type LarkChatListData struct {
	Items     []LarkChat `json:"items,omitempty"`
	Chats     []LarkChat `json:"chats,omitempty"`
	HasMore   bool       `json:"has_more,omitempty"`
	PageToken string     `json:"page_token,omitempty"`
}

// LarkMessagesResponse represents the messages API response
type LarkMessagesResponse struct {
	OK       bool `json:"ok"`
	Data     *LarkMessagesData `json:"data,omitempty"`
	Messages []LarkMessage `json:"messages,omitempty"`
	Error    *LarkError `json:"error,omitempty"`
}

// LarkMessagesData represents the messages data
type LarkMessagesData struct {
	Items     []LarkMessage `json:"items,omitempty"`
	Messages  []LarkMessage `json:"messages,omitempty"`
	HasMore   bool          `json:"has_more,omitempty"`
	PageToken string        `json:"page_token,omitempty"`
}

// LarkError represents a Lark API error
type LarkError struct {
	Type    string `json:"type"`
	Message string `json:"message"`
}

// LarkDoctorResponse represents the doctor command response
type LarkDoctorResponse struct {
	OK     bool              `json:"ok,omitempty"`
	Checks []LarkDoctorCheck `json:"checks,omitempty"`
}

// LarkDoctorCheck represents a single doctor check
type LarkDoctorCheck struct {
	Name    string `json:"name"`
	Status  string `json:"status"`
	Message string `json:"message,omitempty"`
}

// LarkChatsResponse represents the enriched chats API response
type LarkChatsResponse struct {
	OK     bool             `json:"ok"`
	Chats  []LarkChatItem   `json:"chats"`
	Filter LarkChatFilter   `json:"filter"`
}

// LarkChatItem represents an enriched chat item
type LarkChatItem struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	MemberCount  int    `json:"member_count"`
	Filtered     bool   `json:"filtered"`
}

// SyncRequest represents a sync request
type SyncRequest struct {
	ChatID   string `json:"chat_id,omitempty"`
	DaysBack int    `json:"days_back,omitempty"`
	Stream   bool   `json:"stream,omitempty"`
}

// SyncResult represents a single chat sync result
type SyncResult struct {
	Inserted int    `json:"inserted"`
	Skipped  int    `json:"skipped"`
	Error    string `json:"error"`
}

// SyncResponse represents the sync API response
type SyncResponse struct {
	OK     bool                  `json:"ok"`
	Error  string                `json:"error,omitempty"`
	Synced map[string]SyncResult `json:"synced"`
}

// SyncProgressEvent represents a sync progress SSE event
type SyncProgressEvent struct {
	Type    string `json:"type"`
	ChatID  string `json:"chatId,omitempty"`
	Phase   string `json:"phase,omitempty"`
	Count   int    `json:"count,omitempty"`
	OK      bool   `json:"ok,omitempty"`
	Synced  map[string]SyncResult `json:"synced,omitempty"`
	Error   string `json:"error,omitempty"`
}
